compiler/noirc_frontend/src/elaborator/comptime.rs

```diff
@@ -63,4 +78,320 @@ impl<'context> Elaborator<'context> {
+        &mut self,
+        attributes: &[SecondaryAttribute],
+        item: Value,
+        span: Span,
+        generated_items: &mut CollectedItems,
+    ) {
+        for attribute in attributes {
+            if let SecondaryAttribute::Custom(name) = attribute {
+                if let Err(error) =
+                    self.run_comptime_attribute_on_item(name, item.clone(), span, generated_items)
+                {
+                    self.errors.push(error);
+                }
+            }
+        }
+    }
+
+    fn run_comptime_attribute_on_item(
+        &mut self,
+        attribute: &str,
+        item: Value,
+        span: Span,
+        generated_items: &mut CollectedItems,
+    ) -> Result<(), (CompilationError, FileId)> {
+        let location = Location::new(span, self.file);
+        let Some((function, arguments)) = Self::parse_attribute(attribute) else {
+            // Do not issue an error if the attribute is unknown
+            return Ok(());
+        };
+
+        // Elaborate the function, rolling back any errors generated in case it is unknown
+        let error_count = self.errors.len();
+        let function = self.elaborate_expression(function).0;
+        self.errors.truncate(error_count);
+
+        let definition_id = match self.interner.expression(&function) {
+            HirExpression::Ident(ident, _) => ident.id,
+            _ => return Ok(()),
+        };
+
+        let definition = self.interner.try_definition(definition_id);
+        let Some(DefinitionKind::Function(function)) = definition.map(|d| &d.kind) else {
+            return Err((ResolverError::NonFunctionInAnnotation { span }.into(), self.file));
+        };
+
+        let function = *function;
+        let mut interpreter = self.setup_interpreter();
+        let mut arguments =
+            Self::handle_attribute_arguments(&mut interpreter, function, arguments, location)
+                .map_err(|error| {
+                    let file = error.get_location().file;
+                    (error.into(), file)
+                })?;
+
+        arguments.insert(0, (item, location));
+
+        let value = interpreter
+            .call_function(function, arguments, TypeBindings::new(), location)
+            .map_err(|error| error.into_compilation_error_pair())?;
+
+        if value != Value::Unit {
+            let items = value
+                .into_top_level_items(location)
+                .map_err(|error| error.into_compilation_error_pair())?;
+
+            self.add_items(items, generated_items, location);
+        }
+
+        Ok(())
+    }
+
+    /// Parses an attribute in the form of a function call (e.g. `#[foo(a b, c d)]`) into
+    /// the function and quoted arguments called (e.g. `("foo", vec![(a b, location), (c d, location)])`)
+    fn parse_attribute(annotation: &str) -> Option<(Expression, Vec<Expression>)> {
+        let (tokens, lexing_errors) = Lexer::lex(annotation);
+        if !lexing_errors.is_empty() {
```

Comment:
If a lexer error occurs, will the annotation silently fail or is there somewhere else where the error will propagate from?

Alternatively, _can_ a lexer error occur here? I'm not sure, but if the lexer sees the same string originally then I'd expect it to always succeed.

Reply:
I was trying to maintain backwards compatibility with custom attributes before but I don't think anyone is using attributes which fail to lex entirely. I've added some errors here. I've kept no errors in the case that the attribute fails to parse into a variable or function in case those are used. E.g. `#[my_attr(foo = bar)]` would fail to parse.